dev-docs/batch-system.rst

```diff
@@ -0,0 +1,784 @@
+required. It then spins up a maximum of 10 instances each time the
+autoscaler runs to avoid cloud provider API rate limits. This approach
+works well for large workloads that have long running jobs. It is not
+very efficient if there's many short running jobs and the driver
+cannot handle the load from a large cluster or the workload is large
+but runs quickly.
+
+Due to differences in resource prices across regions and extra fees
+for inter-region data transfer, the autoscaler needs to be aware of
+the regions a job can run in when scaling up the cluster in order to
+avoid suboptimal cluster utilization or jobs not being able to be
+scheduled due to a lack of resources.
+
+The current autoscaler works by running every 15 seconds and executing
+the following operations to determine the optimal number of instances
+to spin up per region:
+
+1. Get the fair share resource allocations for each user across all
+   regions and figure out the share for each user out of 300 (this
+   represents number of scheduling opportunities this user gets
+   relative to other users).
+2. For every user, sort the "Ready" jobs by regions the job can run in
+   and take the first N jobs where N is equal to the user share
+   computed in (1) multiplied by the autoscaler window, which is
+   currently set to 2.5 minutes. The logic behind this number is it
+   takes ~2.5 minutes to spin up a new instance so we only want to
+   look at a small window at a time to avoid spinning up too many
+   instances. It also makes this query feasible to set a limit on it
+   and only look at the head of the job queue.
+3. Take the union of the result sets for all of the users in (2) in
+   fair share order. Do another pass over the result set where we
+   assign each job a scheduling iteration which represents an estimate
+   of which iteration of the scheduler that job will be scheduled in
+   assuming the user's fair share.
+4. Sort the result set by user fair share and the scheduling iteration
+   and the regions that job can run in. Aggregate the free cores by
+   regions in order in the result set. This becomes the number of free
+   cores to use when computing the number of required instances and
+   the possible regions the instance can be spun up in.
+
+
+*********
+Scheduler
+*********
+
+The scheduler finds the set of jobs to schedule by iterating through
+each user in fair share order and then scheduling jobs with a "Ready"
+state until the user's fair share allocation has been met. The result
+set for each user is sorted by regions so that the scheduler matches
+what the autoscaler is trying to provision for. The logic behind
+scheduling is not very sophisticated so it is possible to have a job
+get stuck if for example it requires 8 cores, but two instances are
+live with 4 cores each.
+
+Once the scheduler has assigned jobs to their respective instances, in
+groups of 50, the scheduler performs the work necessary to grab any
```

Comment:
Is groups of 50 referring to grabbing secrets or assigning jobs to instances?

Reply:
I thought the worker asked the driver to read the secrets for it; how is the scheduler code involved in secret reading?

Reply:
the scheduler calls `schedule_job` which prepares the job config with the secrets filled in that is then sent to the worker.